src/core/PasswordHealth.h

```diff
@@ -0,0 +1,107 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef KEEPASSX_PASSWORDHEALTH_H
+#define KEEPASSX_PASSWORDHEALTH_H
+
+#include <QHash>
+#include <QSharedPointer>
+#include <QStringList>
+
+class Database;
+class Entry;
+class QString;
+
+class PasswordHealth
```

Comment:
I have a lot of concerns with the way this class was implemented, chief among them is the way the cache is handled. There is also inconsistent use of pointers and references when compared to the rest of the code base. I can fix this one to comply. There are 5 constructors and each of them does something vastly different than the others. That is not appropriate and should be handled by member function calls instead. My expectation for usage of this class would be:

```
auto pwHealth = new PasswordHealth(m_db);
for (entry : m_db->entries()) {
	auto score = pwHealth->score(entry);
	....
}
```

All caching and behind the scenes evaluation would be wholly encapsulated within the class. The constructor call with a database would perform an auto-cache of all entry passwords. Then calling evaluate for each entry would do the lookups and scoring as necessary.

Simple evaluations of entropy should be static function calls:
```
auto score = PasswordHealth::evaluate(password);
```

Reply:
You're right, the PasswordHealth class is trying to do two things at a time (manage a single password, and manage all passwords of a database). There should be two classes instead, one that deals with the properties of a password (entropy, color, etc.) and one to manage all the passwords of a database (re-use, expiry, caching, etc.)

Reply:
Refactored it into two classes, `PasswordHealth` for the properties of a single password and `HealthChecker` to deal with a whole database. Your example (`evaluate(password)`) covers the most simple case only because the score may depend on more than just the password; yet, if a password is all you have, it's

```cpp
auto score = PasswordHealth(password).score();
```

If you want the score for an entry (which you need if you want to consider e. g. expiry), you'd do this:

```cpp
auto checker = HealthChecker(db);
auto score = checker(entry).score();
```

The checker object handles all the caching internally (you'd usually create it once and use it for all entries in the database).

Reply:
> There is also inconsistent use of pointers and references when compared to the rest of the code base. I can fix this one to comply.

Do so please. The rules I'm sticking to are "use references if you can, pointers if you must" and "use automatic storage if you can, dynamic storage if you must". Qt conventions (with QSharedPointer etc.) may be different.